pkg/vm/stack.go

```diff
@@ -163,64 +128,43 @@ func NewStack(n string) *Stack {
 
 func newStack(n string, refc *refCounter) *Stack {
 	s := new(Stack)
+	s.elems = make([]Element, 0, 16) // Most of uses are expected to fit into 16 elements.
 	initStack(s, n, refc)
 	return s
 }
 func initStack(s *Stack, n string, refc *refCounter) {
 	s.name = n
 	s.refs = refc
-	s.top.next = &s.top
-	s.top.prev = &s.top
-	s.len = 0
+	s.Clear()
 }
 
 // Clear clears all elements on the stack and set its length to 0.
 func (s *Stack) Clear() {
-	s.top.next = &s.top
-	s.top.prev = &s.top
-	s.len = 0
+	if s.elems != nil {
+		s.elems = s.elems[:0]
+	}
 }
 
 // Len returns the number of elements that are on the stack.
 func (s *Stack) Len() int {
-	return s.len
-}
-
-// insert inserts the element after element (at) on the stack.
-func (s *Stack) insert(e, at *Element) *Element {
-	// If we insert an element that is already popped from this stack,
-	// we need to clean it up, there are still pointers referencing to it.
-	if e.stack == s {
-		e = NewElement(e.value)
-	}
-
-	n := at.next
-	at.next = e
-	e.prev = at
-	e.next = n
-	n.prev = e
-	e.stack = s
-	s.len++
-
-	s.refs.Add(e.value)
-
-	return e
+	return len(s.elems)
 }
 
 // InsertAt inserts the given item (n) deep on the stack.
-// Be very careful using it and _always_ check both e and n before invocation
-// as it will silently do wrong things otherwise.
-func (s *Stack) InsertAt(e *Element, n int) *Element {
-	before := s.Peek(n - 1)
-	if before == nil {
-		return nil
-	}
-	return s.insert(e, before)
+// Be very careful using it and _always_ check n before invocation
+// as it will panic otherwise.
+func (s *Stack) InsertAt(e Element, n int) {
+	s.elems = append(s.elems, e)
+	l := len(s.elems)
+	copy(s.elems[l-n:], s.elems[l-n-1:l-2])
```

Comment:
Are you sure that the source should be ` s.elems[l-n-1:l-2]`? Could you explain, why `l-2`?

Reply:
Otherwise `TUCK` doesn't work. But we can see how it works through example, imagine `TUCK` with 2 elements on the stack, so we do `InsertAt(e, 2)`:
 * (e1, e2) initially, `len == 2`, `n == 2`
 * we do append and get (e1, e2, e), `len == 3` (that's the key!), `n == 2`
 * we then copy from `[3-2-1:1]` to `[3-2:]`, that is from `[0:1]` to `[1:]`, that's what we want to move old (e1, e2) to the top
 * so we get (e1, e1, e2)
 * and then replace element `3-2-1` that is `0` to get (e, e1, e2) which is exactly what we want

Technically, this stack structure opens the door for many other optimizations that can be made. At the moment I've just tried to reuse existing API, so it implements everything it implemented before the change.

Reply:
I agree that it works correctly with `TUCK`, but these changes break `InsertAt` behaviour. Consider the following test:
```

func TestInsertAt(t *testing.T) {
	s := NewStack("stack")
	s.PushVal(1)
	s.PushVal(2)
	s.PushVal(3)
	s.PushVal(4)
	s.PushVal(5)

	e := s.Dup(1) // it's `4`
	s.InsertAt(e, 3)

	assert.Equal(t, int64(5), s.Peek(0).BigInt().Int64()) // <-- succeed on master, but fails here on patch branch
	assert.Equal(t, int64(4), s.Peek(1).BigInt().Int64())
	assert.Equal(t, int64(3), s.Peek(2).BigInt().Int64())
	assert.Equal(t, int64(4), s.Peek(3).BigInt().Int64())
	assert.Equal(t, int64(2), s.Peek(4).BigInt().Int64())
	assert.Equal(t, int64(1), s.Peek(5).BigInt().Int64())
}
```
It passes on master, but fails on `vm-optimize-stack` branch. I would expect from `InsertAt` just to insert the element and do not change the top of the stack. Although we don't use `InsertAt` anywhere except the `TUCK` handler.

So I would suggest to change this line to:
```
copy(s.elems[l-n:], s.elems[l-n-1:l])
```
in order to avoid bugs in future if we're to use `InsertAt` somewhere else.

Reply:
OK, convincing. One thing I've specifically wanted to avoid here is implicit len() use in `[l:]`. Of course practically it doesn't matter, but exactly matching the number of elements moved and capacity of destination slice might still be beneficial. Fixed this and simplified a bit by taking `l` before `append`.